src/analyzer/protocol/smtp/BDAT.cc

```diff
@@ -0,0 +1,301 @@
+#include "zeek/analyzer/protocol/smtp/BDAT.h"
+
+#include "zeek/3rdparty/doctest.h"
+#include "zeek/Conn.h"
+#include "zeek/DebugLogger.h"
+#include "zeek/analyzer/protocol/mime/MIME.h"
+#include "zeek/util.h"
+
+namespace zeek::analyzer::smtp::detail {
+
+
+struct BDATCmd parse_bdat_arg(int length, const char* arg) {
+    const char* arg_end = arg + length;
+    struct BDATCmd r = {0};
+
+    if ( *arg == '\0' || ! isdigit(*arg) ) {
+        r.error = "BDAT not followed by a valid chunk-size";
+        return r;
+    }
+
+    char* chunk_size_end = nullptr;
+    uint64_t chunk_size = strtoul(arg, &chunk_size_end, 10);
+    if ( *chunk_size_end != ' ' && chunk_size_end != arg_end ) {
+        r.error = "BDAT chunk-size not valid";
+        return r;
+    }
+
+    r.chunk_size = chunk_size;
+    r.is_last_chunk = strncasecmp(chunk_size_end, " LAST", 5) == 0;
+    return r;
+}
+
+
+SMTP_BDAT_Analyzer::SMTP_BDAT_Analyzer(Connection* conn, mime::MIME_Message* mail, size_t max_line_length)
+    : analyzer::Analyzer("SMTP_BDAT", conn), max_line_length(max_line_length), mail(mail) {}
+
+void SMTP_BDAT_Analyzer::NextChunk(ChunkType chunk_type, uint64_t chunk_size) {
+    DBG_LOG(DBG_ANALYZER, "BDAT: NextChunk size=%" PRIi64 " last=%d", chunk_size, chunk_type == ChunkType::Last);
+    assert(remaining_chunk_size == 0);
+    cur_chunk_type = chunk_type;
+    remaining_chunk_size = chunk_size;
+}
+
+void SMTP_BDAT_Analyzer::DeliverStream(int len, const u_char* data, bool is_orig) {
+    analyzer::Analyzer::DeliverStream(len, data, is_orig);
+    assert(mail != nullptr);
+    assert(len <= RemainingChunkSize());
```

Comment:
Is it possible to create a case where this triggers?

If yes - this should probably raise a weird instead.

Reply:
Made it a weird and truncated len instead. It shouldn't trigger given the `SMTP_Analyzer` usage, but remember being uneasy having the `assert` back when I added it :-)

Thanks!